### ISSUE TITLE
Use with different video dimensions data

### ISSUE THREAD

**Issue by auesro** [KumarLabJax/MouseTracking#1]
Dear KumarLab,

As you mention in the Docs part: "This software was designed for 480x480 monochromatic images and is untested for different image size. Due to the pooling and upsampling layers, this exact structure will only work with images in multiples of 96 pixels without adjusting the network layers.
Functionally, the input images must be square and be the same shape as all the other images tested."

I would like to use the code on video that is not square (960x540) so I was wondering if you are planning to add the use of videos with different dimensions.

In any case, if you could share some advice on how to modify the network to make it work on other video formats it would be really helpful.

Thanks!

**Comment by SkepticRaven** [comment]
There are 2 options to be able to process the differently shaped videos.

The first is to crop/pad and resize the input video. I typically use ffmpeg to do these operations.
An example crop and resize could be: 
`ffmpeg -i INPUT.avi -vf "crop=540:540:210:0;scale=480:480" -q 0 -c:v mpeg4 CROPPED_OUTPUT.avi`
This will crop the middle of the video out and resize it to 480x480.
An example pad and resize could be: 
`ffmpeg -i INPUT.avi -vf "pad=960:960:0:210;scale=480:480" -q 0 -c:v mpeg4 PADDED_OUTPUT.avi`
This will add black bars on the top/bottom of the video to make it square and then resize it to 480x480.
ffmpeg also has the [perspective](https://ffmpeg.org/ffmpeg-filters.html#perspective) filter, which can be used to select 4 points on the input video to be used as corners of the output video.

If you want to maintain the higher resolution, you could just do a pad without a resize because 960 is divisible by 96. The pre-trained network we supply is trained for 480x480, but you can still train using the --input_size 960 parameter as the reader will attempt to resize the input to that shape. I have not tested inputs other than 480x480, so there might be some unforeseen bugs. Be sure that the input training data is resized to be square, otherwise it will stretch the one dimension until it is square. Since TF1.9, they've added a "resize with crop or pad" [function](https://www.tensorflow.org/versions/r1.9/api_docs/python/tf/image/resize_image_with_crop_or_pad?hl=en), which could also be used to resize in the code instead of using ffmpeg, which could be used to replace the "tf.images.resize_images". The "np.resize" calls should also be altered to match accordingly in the inference code.

The second option is to alter all the code that applies this restriction. I've used 1 number for both the input width and height, which would need extensive changes across multiple files. Wherever "input_size" is used to describe both image height and width would need to be changed. Also, training augmentation for random 90 deg rotations would need to be changed, since we use the faster transpose operations (instead of an actual rotation). We currently don't have plans to make these changes, but encourage you to fork the project and make the changes yourself if you feel this is the better solution.

**Comment by auesro** [comment]
Thanks for your reply, Brian.
My videos need to be encoded anyway (just a *.h264 stream was captured) and I am doing that with ffmpeg so adding the padding step seems like the easiest approach and should be straightforward, I will give it a try.
I dont think I have the skills to modify the whole code, but if I get some free time I might play with it.

**Comment by auesro** [comment]
Hi Brian,
I might have missed something.
I started with the Labeler and while it opens and let me mark foreground, backwround and apply the watershed algorithm, it does not respond to the arrow keys, so I cant go to the next image...any suggestions?

**Comment by SkepticRaven** [comment]
I've attempted to make it cross-platform as best as possible, but the arrow keys get different signals based on the operating system. The keys might be the numpad directions (4 = left, 8 = up, 6 = right, 2 = down). I switched to these when the system I tested on would inconsistently respond to the arrow keystrokes.

If you're using windows, I have alternative keys commented out in the [code](https://github.com/KumarLabJax/MouseTrackingExtras/blob/b32500aa16dbbb73952ad5c79b30edbbc8322637/Watershed_Labeler/Watershed_Interactive_Release.cpp#L41) that might work.

If none of those work, we'll have to find the keystroke values that your system is producing.

**Comment by auesro** [comment]
I see
Right now im on Linux Mint and I have got this:
Key 2: Angle 330
Key 4: Angle 330
Key 6: Angle 150
Key 8: Angle 150
Key N: skip frame
But I m missing the other 2 angle possibilities...

**Comment by SkepticRaven** [comment]
That's functioning as intended.
While you can select in 90-deg intervals, the exact ellipse angle is determined by the predicted mask. The keystrokes will only select one of two angles, one being 180 deg offset.

Without selecting a mask, the default angle is either up or down.

**Comment by auesro** [comment]
Ok, got it working then.
Only that in my system I need to use the strip of numbers above the alphabet letters, doesnt work with the numpad keys.
Thanks!

**Comment by auesro** [comment]
Ok, I tried to run the training (I just labeled a few images, around 100) to see if everything works. 
Padded the videos to 960x960 as suggested, and changed the input size to 960:

```
(Labeler) auesro@auesro-Mint:~/Desktop/MouseTracking-master$ python main.py --net_type segellreg --batch_size 50 --input_size 960 Train --model construct_segellreg_v8 --log_dir Training_Segellreg --learn_function gen_train_op_adam --train_list xaa --valid_list xab --start_learn_rate 1e-5
Constructing model...
/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/scipy/ndimage/morphology.py:113: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  out[slc] = structure != 0
Adding loss function...
WARNING:tensorflow:From /home/auesro/Desktop/MouseTracking-master/utils/training.py:10: mean_squared_error (from tensorflow.contrib.losses.python.losses.loss_ops) is deprecated and will be removed after 2016-12-30.
Instructions for updating:
Use tf.losses.mean_squared_error instead.
WARNING:tensorflow:From /home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/contrib/losses/python/losses/loss_ops.py:539: compute_weighted_loss (from tensorflow.contrib.losses.python.losses.loss_ops) is deprecated and will be removed after 2016-12-30.
Instructions for updating:
Use tf.losses.compute_weighted_loss instead.
WARNING:tensorflow:From /home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/contrib/losses/python/losses/loss_ops.py:146: add_arg_scope.<locals>.func_with_args (from tensorflow.contrib.losses.python.losses.loss_ops) is deprecated and will be removed after 2016-12-30.
Instructions for updating:
Use tf.losses.add_loss instead.
Populating input queues...
Starting input threads...
Initializing optimizer...
Generating summaries and savers...
Initializing model...
Beginning training...
Traceback (most recent call last):
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1322, in _do_call
    return fn(*args)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1307, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1409, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor of shape [50,480,480,16] and type float
	 [[Node: Optimizer/gradients/zeros_56 = Const[dtype=DT_FLOAT, value=Tensor<type: float shape: [50,480,480,16] values: [[[0 0 0]]]...>, _device="/job:localhost/replica:0/task:0/device:GPU:0"](^ConstantFoldingCtrl/Optimizer/gradients/Switch_58_1)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 101, in <module>
    main(sys.argv[1:])
  File "main.py", line 92, in main
    trainNet.trainNetwork(arg_dict)
  File "/home/auesro/Desktop/MouseTracking-master/utils/TrainNetwork.py", line 277, in trainNetwork
    trainSegEllfitNetwork(arg_dict)
  File "/home/auesro/Desktop/MouseTracking-master/utils/TrainNetwork.py", line 131, in trainSegEllfitNetwork
    _, train_loss, summary_output, cur_step, ellfit_errs = sess.run(fetches=[train_op, loss, training_summary, global_step, errors2], feed_dict={image_placeholder: img_batch, seg_label_placeholder: seg_label_batch, ellfit_label_placeholder: ellfit_label_batch, is_training: True})
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 900, in run
    run_metadata_ptr)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1135, in _run
    feed_dict_tensor, options, run_metadata)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1316, in _do_run
    run_metadata)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1335, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor of shape [50,480,480,16] and type float
	 [[Node: Optimizer/gradients/zeros_56 = Const[dtype=DT_FLOAT, value=Tensor<type: float shape: [50,480,480,16] values: [[[0 0 0]]]...>, _device="/job:localhost/replica:0/task:0/device:GPU:0"](^ConstantFoldingCtrl/Optimizer/gradients/Switch_58_1)]]

Caused by op 'Optimizer/gradients/zeros_56', defined at:
  File "main.py", line 101, in <module>
    main(sys.argv[1:])
  File "main.py", line 92, in main
    trainNet.trainNetwork(arg_dict)
  File "/home/auesro/Desktop/MouseTracking-master/utils/TrainNetwork.py", line 277, in trainNetwork
    trainSegEllfitNetwork(arg_dict)
  File "/home/auesro/Desktop/MouseTracking-master/utils/TrainNetwork.py", line 112, in trainSegEllfitNetwork
    learn_rate, train_op = arg_dict['learn_function'](loss, arg_dict['dataset'].train_size, arg_dict['batch_size'], global_step, arg_dict['start_learn_rate'], arg_dict['epocs_per_lr_decay'], const_learn_rate=arg_dict['const_learn_rate'])
  File "/home/auesro/Desktop/MouseTracking-master/utils/training.py", line 134, in gen_train_op_adam
    train_op = slim.learning.create_train_op(loss, optimizer)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/contrib/slim/python/slim/learning.py", line 439, in create_train_op
    check_numerics=check_numerics)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/contrib/training/python/training/training.py", line 447, in create_train_op
    colocate_gradients_with_ops=colocate_gradients_with_ops)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 526, in compute_gradients
    colocate_gradients_with_ops=colocate_gradients_with_ops)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 494, in gradients
    gate_gradients, aggregation_method, stop_gradients)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 625, in _GradientsHelper
    out_grads[i] = loop_state.ZerosLike(op, i)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/ops/control_flow_ops.py", line 1352, in ZerosLike
    return ZerosLikeOutsideLoop(op, index)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/ops/control_flow_ops.py", line 1476, in ZerosLikeOutsideLoop
    return array_ops.zeros(zeros_shape, dtype=val.dtype)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py", line 1626, in zeros
    output = fill(shape, constant(zero, dtype=dtype), name=name)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 2717, in fill
    "Fill", dims=dims, value=value, name=name)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3392, in create_op
    op_def=op_def)
  File "/home/auesro/anaconda3/envs/Labeler/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1718, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

ResourceExhaustedError (see above for traceback): OOM when allocating tensor of shape [50,480,480,16] and type float
	 [[Node: Optimizer/gradients/zeros_56 = Const[dtype=DT_FLOAT, value=Tensor<type: float shape: [50,480,480,16] values: [[[0 0 0]]]...>, _device="/job:localhost/replica:0/task:0/device:GPU:0"](^ConstantFoldingCtrl/Optimizer/gradients/Switch_58_1)]]

(Labeler) auesro@auesro-Mint:~/Desktop/MouseTracking-master$ 
```

It doenst look good, does it?

**Comment by SkepticRaven** [comment]
That's an OOM error. The network is too big to actually compute on the hardware you have. The 50-batch size was used with 480x480 input and a GPU with 12GB of RAM.

An easy option would be to reduce the batch size (since that reduces the memory footprint). Reducing it to 5 or 10 will help, but it might not be enough. You might not want to go much smaller than 4 because there's a handful of papers that have issues with tiny batches.
You could also reduce the width of the network by changing [this number](https://github.com/KumarLabJax/MouseTracking/blob/30acc45e94adcbc8ea0ae3ea840c434e6ce11aab/utils/models.py#L80).

Other options are a bit less ideal. Resizing the input image. Adding more RAM to your system. Changing the [optimizer](https://github.com/KumarLabJax/MouseTracking/blob/30acc45e94adcbc8ea0ae3ea840c434e6ce11aab/utils/training.py#L132) from adam to the less memory intensive [SGD optimizer](https://www.tensorflow.org/api_docs/python/tf/train/GradientDescentOptimizer).

**Comment by auesro** [comment]
Oh, thats good news. 
Im playing with my system at home (GPU with 8GB RAM), I can go bigger at the lab.
I just padded+resized my video to 480x480 and labelled some frames, and its running using batch_size 10 and using your fullmodel as network_to_restore.
GPU is engaged and training is proceeding. How long do you recommend I train for?

**Comment by SkepticRaven** [comment]
It all depends on the data and learning rate. The rule of thumb is you wait until the validation curve starts to level off (inferring that passing the data through again doesn't add performance). If it starts doing worse, that means the network is memorizing the training dataset.

I typically set it for 500k steps, but most of my trainings are done around 100-200k step.

**Comment by auesro** [comment]
Ok (and excuse my ignorance on this department) you mean when the "loss" parameter that its being constantly updated in Terminal stays stable, the training is done, right? I can stop at any time with ctrl+c?

**Comment by auesro** [comment]
So I used ctrl+c to stop the training.
I decided to test now with my 960x960 images:
-trying to continue training from your fullmodel gave a tensorflow error indicating something about things not being the same shape so...yea, no luck
-so I decided to start from scratch using batch_size 10 and so far it is working (step 280 now).
My doubt (together with the one regarding loss from the previous comment) is...do xerr and yerr give the error in pixels for the ellipse centroid? I guess they should go down together with the loss, right? Because so far they are almost as big as the frame (around 600-700).

**Comment by SkepticRaven** [comment]
Yes, you want to watch the loss value. The terminal output is for the training loss. That's a gauge of "if it can do better on the training set". It won't tell you if it's memorizing the training dataset.

To watch the validation and training losses at the same time... There should be an "events" file that is created in the folder where the newly trained model is saved. This is a tensorboard file that makes visualization pretty easy.
`tensorboard --logdir <folder_with_that_file>`

As for the big errors (yes, they're in pixels), that's mostly an artifact of my approach of setting negative default values when no mask is produced. They'll jump down to more reasonable numbers as the network learns enough to predict that an object is actually in the image.

**Comment by auesro** [comment]
This looks great!
I trained your Full Model on 60 of my labelled images for 6500 steps and inferred an almost perfect fit ellipse for my mouse, amazing!
However, I selected --ellfit_movie_output --ellfit_features_output --ellfit_output and the only file it generated (apart from the movie) was an *.npz file...any ideas on this?

Also, on the other hand, I had a look at your files searching for those "input_size" parameters that would need to be changed in order to use non-squared videos. Excuse my ignorance, but would it be enough to change the type from int to tuple for all of it to work? If Im totally lost, just let me know!

**Comment by auesro** [comment]
Ok, so I have been tackling the issue of different video dimensions. I went through the input_sizes and changed them accordingly to take the "new" input size list [width, height].
That went well, I can run the training on square videos after such modifications.
However, when trying to train the network on non-squared videos, I face the problem that the Encoder-Decoder design, generates non-integer values for the height dimension (540) on p3 (models.py, construct_segellreg_v8). I have tried to modify the stride and the padding but unsuccesfully so far...any ideas to overcome this?

**Comment by SkepticRaven** [comment]
The npz file is an n-by-6 numpy array, written in append mode (which technically isn't really supported by numpy). I've written an "export to csv" python function over [here](https://github.com/KumarLabJax/MouseTrackingExtras/tree/master/NPYReader).

As for the non-integers in the network... you'll need to define a new shape that the input is divisible by.
The shapes that are important to look at are the slim.max_pool2d and slim.conv2d_transpose. Specifically in them, the strides are what alters the shape. I do 5 2x2 pools followed by a 3x3 pool for a total of 96x reduction. The transpose convs then reverse them. Since you have 540 and 960, the least common multiple is 2^2 * 3^3 * 5^1.  You could use 2 2x2 and 3 3x3.

**Comment by auesro** [comment]
Thanks for the info.
Yes, I have used already your export to csv function, no problem there.

After quite a bit of reading, I arrived to the same conclusion you are proposing.  But my idea was to make it work for any kind of video dimensions which seems pretty tough (from my lack of experience with convnets, that is).

For now, I have figured a way to input non-padded videos using the "resize with crop or pad" function you suggested from TF (already available in TF 1.8, which I am using) to pad them in-code. And also how to modify the inference code to make it work.

Right now, the last thing I need is to understand the ellfit_output. I need to modify the code so that I get the angle of the ellipse in radians, growing counterclock-wise and zero pinting right (east), because thats how JAABA takes the angles. I have kind of succeded but the sine and cosine values in ellfit.npz (last 2 columns) seem odd to me. From what I understand, what we get in ellfit.npz are the sine and cosine of the angle value (in radians) transformed to degrees ([here](https://github.com/KumarLabJax/MouseTracking/blob/30acc45e94adcbc8ea0ae3ea840c434e6ce11aab/utils/readers.py#L49)), am I right? 
If thats correct, I only need to calculate the arcsin/arcos, transform it to radians and "move" it by 3pi/2, did I get it?
Thanks a lot for your help, Brian!

**Comment by SkepticRaven** [comment]
I typically use an arctan2 function to pull out the angle. Going from network output to radians should be:
`np.arctan2(network_output[:,5],network_output[:,4])*180/np.pi`

**Comment by auesro** [comment]
Yes, I found the way you do it with arctan2 in plotters.py
That gives me the angle in degrees, but with negative values when it goes beyond 180, right? I mean considering 0 the downward pointing direction, it gives positives values for the "right" half of the circle but negative values for the "left" half of the circle, right? It does not come in 0-360 degrees range

**Comment by SkepticRaven** [comment]
The numpy arctan2 function returns -pi to pi. If you want to convert to 0-360 range with 0 being east...
`angle = (np.arctan2(network_output[:,5],network_output[:,4])*180/np.pi + 90)%360`
+90 deg shift to go from south to east. Modulo 360 to give in 0-360 range (corrects the negative numbers nicely).

**Comment by auesro** [comment]
Man! Thats what I was looking for! Thanks a lot! I spent yesterday a few hours trying to figure out a way to get rid of the negative values...do you know if you can apply the same in radians? Something like:
angle = (np.arctan2(network_output[:,5],network_output[:,4]) + np.pi/2)%2np.pi

np.pi/2 is 90 degrees in radians
2np.pi is 360 degrees in radians...not sure how it will interact with %

**Comment by SkepticRaven** [comment]
I'd recommend parenthesis around the 2*np.pi. Modulo and multiply are same precedence so without it you'll be doing modulo 2 then mulitply by pi. Otherwise, should be fine (decimals work with modulo).

**Comment by auesro** [comment]
Unfortunately, it does not seem to work:
- On the one hand, if put between parenthesis, modulo ignores the content of the parenthesis.

- On the other hand, and more importantly, the angle I get from the formula do not correspond with what I see in the movie output...At first, I thought the calculation should be:
`angle = (np.arctan2(network_output[:,4],network_output[:,5])*180/np.pi + 90)%360`
since [4] is the sine and [5] is the cosine (according to [this](https://github.com/KumarLabJax/MouseTracking/blob/30acc45e94adcbc8ea0ae3ea840c434e6ce11aab/utils/readers.py#L49) line) and the mathematical formula says that [theta = arctan(sin/cosine)](https://en.wikipedia.org/wiki/Inverse_trigonometric_functions#In_computer_science_and_engineering). Then arctan2 [documentation](https://docs.scipy.org/doc/numpy-1.14.5/reference/generated/numpy.arctan2.html) says that the "y coordinate" is the first argument of the function, so kind of confirming that the sine value should be first in the angle calculation. Even doing this, still I do not get values that could possibly fit with what I see in the corresponding frame...
So one question, when you write this:
`labelAngle = -np.arctan2(label[4],label[5])*180/np.pi` in [plotters.py](https://github.com/KumarLabJax/MouseTracking/blob/30acc45e94adcbc8ea0ae3ea840c434e6ce11aab/utils/plotters.py#L8), is this the code that calculates the angle with which to plot the line inside the ellipse in the videos generated by the --ellfit_movie_output? If so...any ideas to solve my issue? (reminder: I just need to modify the ellfit.npy file to have on the 5th column the angle of the ellipse in order to feed it to JAABA).

**Comment by SkepticRaven** [comment]
Yes, the line in plotters.py is what is used in the plotting for inference videos. It's using the opencv function for the ellipse (which has the convention of 0 being east, the same as your intent).
No, I don't use the angle to plot the _line_ (I actually just use the already calculated sin/cos values). Although [that code](https://github.com/KumarLabJax/MouseTracking/blob/30acc45e94adcbc8ea0ae3ea840c434e6ce11aab/utils/plotters.py#L13) seems to be confusing sin/cos as well.

It's been awhile since I've looked at that code, so let's use some simple example to figure out what's going on. I recall having difficulties getting it to plot correctly probably due to differences between opencv, Ctrax, and a handful of other software I was using. The current code produces visually correct results.

```
angles = np.array([45, 90, 135, 180, 270])
temp = [[np.sin(np.pi/180*x), np.cos(np.pi/180*x)] for x in angles] # note that sin is [0] and cos is [1]

# reproduces original angles
np.array([np.arctan2(x[0], x[1])*180/np.pi for x in temp])%360
# array([ 45.,  90., 135., 180., 270.])

approach1 = np.array([np.arctan2(x[0], x[1])*180/np.pi + 90 for x in temp])
approach2 = np.array([-np.arctan2(x[1], x[0])*180/np.pi for x in temp])

# what new angles should be...
angles + 90
# array([135, 180, 225, 270, 360])
approach1%360
# array([135., 180., 225., 270., 360.])
approach2%360
# array([315., 360.,  45.,  90., 180.])
(approach2+180)%360
# array([135., 180., 225., 270.,   0.])
```
I'm pretty sure Ctrax inverts the y-axis which caused me some confusion. I'm wondering if JAABA does the same and that could be another source of confusion.

**Comment by auesro** [comment]
Ok, now everything seems to work using the formula:
`angle = (np.arctan2(network_output[:,4],network_output[:,5])*180/np.pi + 90)%360`
This code produces visually correct results as you say. Except that summing 90 degrees does not change the 0 reference. Lets see if I can explain it:
-angle 35 with reference south (as provided by your code in the ellfit file)
-35 +90 degrees is 125 degrees. Thats not the same position of the animal in the east reference space. The same position would be 35 + 270= 305 degrees. So it does not work summing 90 degrees to all the values provided by the code, but it seems to work with 270. 
So the code should be, in my case:
`angle = (np.arctan2(network_output[:,4],network_output[:,5])*180/np.pi + 270)%360`
And then the modulo takes care of the values higher than 360. Am I right?

**Comment by SkepticRaven** [comment]
Yes, modulo forces the range of output to be between 0 and the modulo operand. In this case, 0-360.

Also, that does appear to be a goofup on the direction of the shift in my initial recommendation. Going from a south to east reference point should have been _subtracting_ 90. Going from east to south reference point is adding 90. Adding 270 is equivalent to subtracting 90.

**Comment by auesro** [comment]
Hi again Brian,
I have it all set up now and working except for one little issue which might or might not be related to the Tracker itself:
I get both the ellfit.npz and ellfit_movie outputs with 90450 frames/rows while the original movie to be inferred is 90479 frames (calculated with imageio and with opencv, in 2 different computers). So apparently it is missing the last 29 frames. I set a batch size of 50, so the 90450 makes sense, it looks as if the last batch (last 29 frames) could not be read since at the end of the Inferring process I get the following message:
`UserWarning: Could not read last frame of /home/auesro/Desktop/MouseTracking-master_AER_pad/Tests/A.mp4.
  warn('Could not read last frame of %s.' % uri)`
Screenshot:
![Screenshot from 2019-05-19 12-14-36](https://user-images.githubusercontent.com/24290290/57982131-77083400-7a41-11e9-83b8-4a32e656c0c2.png)

Screenshot comparing the size of the original movie and the movie generated by the code:
![Screenshot from 2019-05-19 14-30-51](https://user-images.githubusercontent.com/24290290/57982242-ca2eb680-7a42-11e9-9e00-2da3ffca0085.png)


Have you ever tested movies whose total number of frames is not exactly divisible by your batch_size?

**Comment by SkepticRaven** [comment]
Yeah, the inference code currently only processes data divisible by the batch size. I've been wanting to do a rewrite of the inference code to fix this (and add threading), but haven't found the time.
The easy solution is to do a batch size of 1, because then all frames will be processed. Depending on the GPU you use, there isn't much of a performance change on the processing throughput.

In our lab, we overshoot on the duration of the videos so the not processed frames aren't particularly needed (1 hour tests have 1hr5m recordings). That being said, using a batch size of 50 can only ever not process 1.6s worth of data.

**Comment by auesro** [comment]
Yea, I figured it had to be related.
I have got an RTX2080Ti in the lab so it should run no problem.
Its true, I dont mind about those 1.6 seconds of data but JAABA does...it requires the video file and track file to be exactly the same length so...I will go the batch_size 1 route.

Now I am installing everything in the lab computer and facing a new problem. Maybe I should open a different issue..let me know if you prefer that:

Running the Labeler code it does not save anything...it works as intended, the Terminal displays the "Frame Num , Frame Saved: , Angle" correctly but the Ell, Seg, Reg folders are empty...
The only difference between both machines is CUDA. CUDA 9 at home, CUDA 10 in the lab...but I doubt this little program is using CUDA right? Any ideas?
